domain/consensus/model/staging_area.go

```diff
@@ -0,0 +1,85 @@
+package model
+
+import "github.com/pkg/errors"
+
+// StagingShard is an interface that enables every store to have it's own Commit logic
+// See StagingArea for more details
+type StagingShard interface {
+	Commit(dbTx DBTransaction) error
+}
+
+// StagingShardID is used to identify each of the store's staging shards
+type StagingShardID byte
+
+// StagingShardID constants
+const (
+	StagingShardIDAcceptanceData StagingShardID = iota
+	StagingShardIDBlockHeader
+	StagingShardIDBlockRelation
+	StagingShardIDBlockStatus
+	StagingShardIDBlock
+	StagingShardIDConsensusState
+	StagingShardIDDAABlocks
+	StagingShardIDFinality
+	StagingShardIDGHOSTDAG
+	StagingShardIDHeadersSelectedChain
+	StagingShardIDHeadersSelectedTip
+	StagingShardIDMultiset
+	StagingShardIDPruning
+	StagingShardIDReachabilityData
+	StagingShardIDUTXODiff
+	// Always leave StagingShardIDLen as the last constant
+	StagingShardIDLen
+)
+
+// StagingArea is single changeset inside the consensus database, similar to a transaction in a classic database.
+// Each StagingArea consists of multiple StagingShards, one for each dataStore that has any changes within it.
+// To enable maximum flexibility for all stores, each has to define it's own Commit method, and pass it to the
+// StagingArea through the relevant StagingShard.
+//
+// When the StagingArea is being Committed, it goes over all it's shards, and commits those one-by-one.
+// Since Commit happens in a DatabaseTransaction, a StagingArea is atomic.
+type StagingArea struct {
+	shards     [StagingShardIDLen]StagingShard
+	isCommited bool
+}
+
+// NewStagingArea creates a new, empty staging area.
+func NewStagingArea() *StagingArea {
+	return &StagingArea{
+		shards:     [StagingShardIDLen]StagingShard{},
+		isCommited: false,
+	}
+}
+
+// GetOrCreateShard attempts to retrieve a shard with the given name.
+// If it does not exist - a new shard is created using `createFunc`.
+func (sa *StagingArea) GetOrCreateShard(shardID StagingShardID, createFunc func() StagingShard) StagingShard {
```

Comment:
Please note sa.shards is an array.
This was done deliberately, because map access is much slower (I've tested this with integers, I'd expect the same to happen with pointers, because those are basically integers as well).

This, as you mentioned, has some worse properties, but using a map adds around 20~30 overhead for kaspad normal operation.

This also explains why in Commit a shard might be nil - we are going over all the array every time we Commit. Some slots in the array could be empty.

I've added some comments explaining this, please LMK if there's anywhere else you think should be augmented, or if you have some ideas to how we can improve on this.